add/remove VDPA devices

lmilleri opened the pull request:
This PR extends the govdpa library by adding the following capabilities:

- kvdpa-cli: Add support for new VdpaDevice
        This change allows the user (cli) to create a new VDPA device
        CLI command: kvdpa-cli add [mgmtdev] [dev]
             mgmtdev: the VDPA management device (e.g. pci/0000:65:00.2)
             dev:          the VDPA device (e.g. vdpa1)

- kvdpa-cli: Delete support for VdpaDevice
        This change allows the user (cli) to delete a VDPA device
        CLI command: kvdpa-cli del [dev]
             dev:     the VDPA device (e.g. vdpa1)

- Added function for returning the VdpaDevice objects for a given pciAddress
         function name: GetVdpaDevicesByPciAddress
         arguments: pciAddress string
         return values: []VdpaDevice, error

lmilleri commented:
@amorenoz please take a look

amorenoz commented on cmd/kvdpa-cli/kvdpa-cli.go:
```diff
@@ -72,6 +73,19 @@ func getAction(c *cli.Context) error {
 
+func addAction(c *cli.Context) error {
+	if c.Args().Len() != 2 {
+		usage := fmt.Sprintf("Usage: %v %v %v \n", c.App.Name, c.Command.Name, c.Command.ArgsUsage)
```
This is not very maintainable. You can use cli.ShowAppHelp(c) to show the auto-generated help.

lmilleri replied:
done

amorenoz commented on cmd/kvdpa-cli/kvdpa-cli.go:
```diff
@@ -72,6 +73,19 @@ func getAction(c *cli.Context) error {
+	mgmtDevName := c.Args().Get(0)
+	devName := c.Args().Get(1)
+
+	_, err := vdpa.AddVdpaDevice(mgmtDevName, devName)
```
Adding this commit before actually adding `AddVdpaDevice` means this commit does not compile. It would break git bisects or other experiments. It would be much better to sort commits in another way. First adding functionality, then adding clients that use it.

lmilleri replied:
Rebasing again from master fixes the problem

amorenoz commented on pkg/kvdpa/mgmtdev_test.go:
```diff
@@ -187,3 +187,57 @@ func TestMgmtDevGet(t *testing.T) {
+				}),
+				mock.AnythingOfType("[]*nl.RtAttr")).Return([][]byte{}, tt.err)
+
+			_, err := AddVdpaDevice(tt.mgmtDevName, tt.devName)
```
Same as for cmd line. It's better to add unit tests after the code (or at the same time)

amorenoz commented on cmd/kvdpa-cli/kvdpa-cli.go:
```diff
@@ -86,6 +86,18 @@ func addAction(c *cli.Context) error {
 
+func deleteAction(c *cli.Context) error {
+	if c.Args().Len() != 1 {
+		usage := fmt.Sprintf("Usage: %v %v %v \n", c.App.Name, c.Command.Name, c.Command.ArgsUsage)
```
Same as for Add command, use auto-generated usage string

lmilleri replied:
done

amorenoz commented on pkg/kvdpa/device.go:
```diff
@@ -146,17 +146,22 @@ func (vd *vdpaDev) ParentDevicePath() (string, error) {
 	return parent, nil
 }
 
-/* Finds the virtio vdpa device of a vdpa device and returns its path
```
This commit has nothing to do with the commit message

amorenoz commented on pkg/kvdpa/device.go:
```diff
@@ -206,6 +208,16 @@ func GetVdpaDevicesByMgmtDev(busName, devName string) ([]VdpaDevice, error) {
 }
 
+/*GetVdpaDevicesByPciAddress returns the VdpaDevice objects for the given pciAddress */
+func GetVdpaDevicesByPciAddress(pciAddress string) ([]VdpaDevice, error) {
```
The commit message clearly does not represent what's going on here. There are two new functions with clear functionality while the message seems to indicate it's a harmless refactoring.

lmilleri replied:
Rebasing again from master fixes the problem

amorenoz reviewed (commented):
This PR needs rebasing plus a bit of moving things around between the different commits and ordering them so that every commit builds.

lmilleri commented:
I messed up while rebasing from master. I'll cherry-pick my changes and tidy up the commit messages

lmilleri committed:
kvdpa-cli: Add support for new VdpaDevice

This change allows the user (cli) to create a new VDPA device
CLI command: kvdpa-cli add [mgmtdev] [dev]
   mgmtdev: the VDPA management device (e.g. pci/0000:65:00.2)
   dev:     the VDPA device (e.g. vdpa1)

Signed-off-by: Leonardo Milleri <lmilleri@redhat.com>

lmilleri committed:
kvdpa-cli: Delete support for VdpaDevice

This change allows the user (cli) to delete a VDPA device
CLI command: kvdpa-cli del [dev]
	dev:     the VDPA device (e.g. vdpa1)

Signed-off-by: Leonardo Milleri <lmilleri@redhat.com>

lmilleri committed:
Added function GetVdpaDevicesByPciAddress

function name: GetVdpaDevicesByPciAddress
arguments: pciAddress string
return values: []VdpaDevice, error

Signed-off-by: Leonardo Milleri <lmilleri@redhat.com>

lmilleri commented:
@amorenoz Please review it again, I fixed the rebase issue, commit history and few minor changes you requested

amorenoz commented:
I've fixed an issue with golangci-lint. Please rebase to see if tests pass.
Also, nit: It's generally preferred not to add extra commits fixing bugs introduced in the same PR. Code reviews should be integrated (squashed) into the original commits to make a clean history.

lmilleri commented:
Rebased from master and rewritten git history to make it cleaner. Please review the changes

amorenoz commented:
Thanks @lmilleri

amorenoz commented:
/lgtm
